[build] Extract common code and compiler flags

salkinium opened the pull request:
Refactors both entire build script generator modules to pull out the common code and settings into its own separate file and use that get identical compile flags among other things.

Also floating point support to `:build:scons` module. Not sure why I forgot to port this, probably because the DLR SCons tools don't explicitly support it.

cc @fabian-dlr @strongly-typed @rleh @dergraaf

salkinium committed:
[build] Extract common code and compiler flags

This allows for easier sharing of common settings between different
build script generators.

rleh commented on tools/build_script_generator/scons/resources/SConscript.in:
```diff
@@ -55,6 +55,16 @@ env.Append(CPPDEFINES=[
 %% if platform in ["stm32"]
 env["CPU"] = "{{ core }}"
+%% if fp in ["4f", "7f", "7fd"]
+env["FPU_ABI"] = "-mfloat-abi=softfp"
```
Is there a reason not to use `-mfloat-abi=hard`?
https://gcc.gnu.org/onlinedocs/gcc/ARM-Options.html

salkinium replied:
Hm, when I tried some years ago it failed because some library wasn't compiled with the same settings (I think it was newlib), but I can't reproduce this anymore. Let's see if `hard` ABI works in the CI.

rleh commented:
Im my opinion the CMake build scripts should not use different compiler flags different compared to scons, maybe you can fix floating point support there too: https://github.com/modm-io/modm/blob/develop/tools/build_script_generator/cmake/cmake_scripts/configure-stm32-gcc.cmake#L51-L53

salkinium commented:
> CMake build scripts should not use different compiler flags different compared to scons

Yes! I'd like them to be fed from the same source during generation, preferrably with different profiles: {release, debug}. I think we need our own `arm-none-eabi-gcc` SCons tools for that and not the one from DLR tools, which already sets a lot of default flags. cc @dergraaf

chris-durand commented:
The CMake build script has a typo in line 51 ("cortext-m7" instead of "cortex-m7"). Also "mfpu=fpv4-sp-d16" is not an appropriate choice for any cortex-m7 fpu.

salkinium commented:
> The CMake build script has a typo in line 51 ("cortext-m7" instead of "cortex-m7").

Just saw that too.
Hm, I may need to rewrite this whole flags thing, it's pretty messy already. 😢

chris-durand commented:
Also, the CMake file generates a nonsensical makefile without any compiler flags set if the user doesn't explicitly specify `-DCMAKE_BUILD_TYPE=` . Defaulting to release build would be a much better solution.

salkinium committed:
Fix various compilation warnings & errors

salkinium commented:
I'm extending this PR to not just introduce hard floating point ABI but to also use the same compiler flags for both generators.

salkinium commented on repo.lb:
```diff
@@ -48,6 +48,3 @@ def init(repo):
     repo.find_modules_recursive(".", modulefile="*.lb", ignore="repo.lb")
-
-def build(env):
-    env.append_metadata_unique("include_path", "src")
```
Repo can't build, so this doesn't work. @dergraaf

salkinium commented on src/modm/driver/pressure/bmp085.hpp:
```diff
@@ -143,7 +143,7 @@ class Bmp085 : public bmp085, public modm::I2cDevice<I2cMaster, 1>
 public:
 	/// Get the data object for this sensor.
-	inline Data&
+	inline DataBase&
```
Ping @strongly-typed, this fails to compile on GCC 8, due to wrong type. not sure what your intentions were here.

salkinium commented on tools/build_script_generator/scons/resources/SConscript.in:
```diff
@@ -11,6 +11,7 @@
 from os.path import join, abspath
 
 Import("env")
+profile = ARGUMENTS.get("profile", "release")
```
So `scons program profile=debug` would compile and flash the debug version of the program with `-Og` and `MODM_DEBUG_BUILD`.

salkinium committed:
[build] Move `build.{scons, cmake}` to submodules

This allows specifying common options like project name and build path
together and not having to duplicate them for both generators.

salkinium committed:
[build] Adapt examples, tests for :build:* modules

salkinium commented on tools/build_script_generator/cmake/resources/CMakeLists.txt.in:
```diff
@@ -24,11 +24,9 @@ INCLUDE_DIRECTORIES(
+%% for define in flags["CPPDEFINES"]["base"]
 SET_TARGET_PROPERTIES($(CMAKE_PROJECT_NAME).elf PROPERTIES COMPILE_DEFINITIONS "$({{ define }})")
 %% endfor
-%% endif
```
@siy How would I add the debug defines to the debug build in CMake? They're in `flags["CPPDEFINES"]["debug"]`.

In the SCons Builder, I just pass the profile via command line: `scons build profile=debug` which is then chosen via Python:

```py
profile = ARGUMENTS.get("profile", "release")
if profile == "debug":
    env.Append(CPPDEFINES = [
        "MODM_DEBUG_BUILD",
    ])
env["CCFLAGS"] = [
    "-fdata-sections",
   # ...
]
if profile == "release":
    env.Append(CCFLAGS = [
        "-Os",
    ])
if profile == "debug":
    env.Append(CCFLAGS = [
        "-Og",
    ])
```

chris-durand replied:
CMake uses the `CMAKE_BUILD_TYPE` variable to select different build profiles. It is passed to cmake via the `-D` command line option, e.g. `-DCMAKE_BUILD_TYPE=Debug`, similar to the SCons Builder. Depending on CMAKE_BUILD_TYPE different compiler flags will be chosen.

In any case the flags in CMAKE_C_FLAGS/CMAKE_CXX_FLAGS are used. Profile specific flags can be specified in CMAKE_CXX_FLAGS_* (e.g. CMAKE_CXX_FLAGS_DEBUG for Debug). These will be appended to the default flags in CMAKE_CXX_FLAGS.

It is perfectly legal for the user to not specify CMAKE_BUILD_TYPE. In this case the variable is undefined and only the flags in CMAKE_CXX_FLAGS are considered. You'll probably have to set CMAKE_BUILD_TYPE yourself in the CMake file to default to a specific profile (https://cmake.org/pipermail/cmake/2006-February/008371.html).

CMake defines four possible values for CMAKE_BUILD_TYPE: Debug, Release, RelWithDebInfo and MinSizeRel. Others are not possible.

siy replied:
Exactly. I think that it will be enough to set CMAKE_C_FLAGS_DEBUG/CMAKE_CXX_FLAGS_DEBUG. Particular profile then can be selected by user or IDE by passing appropriate command line flag to CMake.

salkinium replied:
Thanks, I've adapted the flag settings, defaulted to release build and added debug defines.

salkinium commented on tools/build_script_generator/common.py:
```diff
@@ -0,0 +1,218 @@
+    return memories
+
+
+def common_compiler_flags(compiler, target, buildlog):
```
These are the compiler flags for modm, I moved some into the debugging profile and merged the AVR and ARM options. I cross-checked most of these with the documentation, added/removed a few.

Comments? @rleh @strongly-typed

chris-durand replied:
@salkinium: Although I'm neither @rleh nor @strongly-typed, I would like to add some comments. If contribution from my side is not desired, feel free too ignore the comments :wink: 

I have just worked through the online documentation of gcc warning options. The first thing I noticed is that some specified flags are already implied by `-Wall -Wextra` (according to gcc docs) and can be removed (at least for gcc versions that can build modm). This could make the list easier to grasp. This applies to:

-  `-Wformat` implied by `-Wall`
-  `-Winit-self` implied by `-Wall`
-  `-Wuninitialized` implied by `-Wall`
-  `-Wunused-parameter` implied by `-Wunused -Wextra`
-  `-Wwrite-strings` is enabled by default 
-  `-Wimplicit` implied by `-Wall`
-  `-Wdelete-non-virtual-dtor` implied by `-Wall`

Are you sure `-fdata-sections` and `-ffunction-sections` are really beneficial? The docs say the following about them:
>Only use these options when there are significant benefits from doing so.  When you specify these options, the assembler and linker create larger object and executable files and are also slower. 

I also found some flags that we could try:
- `-Wduplicated-cond`: Detects duplicate conditions in `if / else if` statements
- `-Wbad-function-cast`: C only, warn on suspicious cast of return values (e.g. int to pointer cast)
- `-Wlogical-op`: Detect suspicious use of logical operators

salkinium replied:
> Although I'm neither @rleh nor @strongly-typed, I would like to add some comments. If contribution from my side is not desired, feel free too ignore the comments 😉

3am brain farts. The correct way would be @modm-io/contributors.  .oO(I should rename that to `@modm-io/milkshake` coz it gets all the boyz to the yard).
Thanks for going throught the loooooong online docs and posting your summary.

> Are you sure -fdata-sections and -ffunction-sections are really beneficial?

Yes, absolutely, since the linker can throw out a lot more functions/data. For the F469/game of life example it's a difference of 12.8kB vs 18.5kB of Flash usage, static RAM usage is almost identical tho. I'd leave it like this.

> I also found some flags that we could try

Added them, they already found a bug.

chris-durand commented on tools/build_script_generator/common.py:
```diff
@@ -0,0 +1,218 @@
+                "-Wdouble-promotion",
+            ]
+        flags["CCFLAGS"]["base"] += [
+            "-Werror=sign-compare",
```
Why is this STM32 specific?

chris-durand commented on tools/build_script_generator/common.py:
```diff
@@ -0,0 +1,218 @@
+            "-mmcu={}".format(target.partname),
+        ]
+        flags["CCFLAGS"]["base"] += [
+            "-Werror=maybe-uninitialized",
```
Why is this specific to avr?

salkinium commented:
I'm happy with this changeset now.

Unfortunately this introduces a breaking change with the build module: `:build.scons` now maps to `:build:scons` (submodule of `:build`), and `:build.cmake` => `:build:cmake`.
This will break RCA code (we always work with HEAD) until I'll update the projects (<1d).
